### PR TITLE
Refactoring project to reactive approach

### PR DESCRIPTION
**Pull request by HNDaG**
Please review changes, Nats part will be added in future commits

### CODE REVIEW

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/PublisherServiceImpl.kt [hunk 1]
```diff
@@ -1,78 +1,133 @@
...
 import com.nikitahohulia.listeningplatform.dto.request.PostDtoRequest
 import com.nikitahohulia.listeningplatform.dto.request.PublisherDtoRequest
 import com.nikitahohulia.listeningplatform.dto.request.toEntity
 import com.nikitahohulia.listeningplatform.dto.response.PostDtoResponse
 import com.nikitahohulia.listeningplatform.dto.response.PublisherDtoResponse
 import com.nikitahohulia.listeningplatform.dto.response.toResponse
+import com.nikitahohulia.listeningplatform.entity.Post
+import com.nikitahohulia.listeningplatform.entity.Publisher
 import com.nikitahohulia.listeningplatform.exception.DuplicateException
+import com.nikitahohulia.listeningplatform.exception.EntityNotFoundException
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/PublisherServiceImpl.kt]: What the difference between `EntityNotFoundException` and `NotFoundException`?

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/repository/CustomPostRepository.kt [hunk 2]
```diff
@@ -2,24 +2,26 @@ package com.nikitahohulia.listeningplatform.repository
...
+    fun deleteByPublisherId(id: ObjectId): Mono<Long>
 
-    fun findAllPostsByCreatorIdOrderByCreatedAt(creatorId: ObjectId): List<Post>
+    fun findAllPostsByCreatorIdOrderByCreatedAt(creatorId: ObjectId): Flux<Post>
 
-    fun findAllPublisherIdByUserId(userId: ObjectId): List<ObjectId>
+    fun findAllPublisherIdByUserId(userId: ObjectId): Flux<ObjectId>
 
-    fun findAllBySubscriptionIds(subscriptionIds: List<ObjectId>, page: Int, size: Int): List<Post>
+    fun findAllBySubscriptionIds(subscriptionIds: Flux<ObjectId>, page: Int, size: Int): Flux<Post>
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/repository/CustomPostRepository.kt:26]:
```suggestion
    fun findAllBySubscriptionIds(subscriptionIds: ObjectId, page: Int, size: Int): Flux<Post>
```

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/repository/CustomPublisherRepositoryImpl.kt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import org.springframework.data.mongodb.core.query.Criteria
+import org.springframework.data.mongodb.core.query.Query
+import org.springframework.stereotype.Repository
+import reactor.core.publisher.Flux
+import reactor.core.publisher.Mono
+
+@Repository
+class CustomPublisherRepositoryImpl(private val mongoTemplate: ReactiveMongoTemplate) : CustomPublisherRepository {
+
+    val collectionName = "publisher"
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/repository/CustomPublisherRepositoryImpl.kt]: this should be moved to `Publisher` class

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/repository/CustomPostRepository.kt (hunk 2 above)
Comment (HNDaG) [reply listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/repository/CustomPostRepository.kt:26]: Here I must have the list, but I will have to this method blocking and even after that still make a flux from iterable, or make this method **private**

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/entity/Post.kt [hunk 4]
```diff
@@ -11,6 +11,10 @@ data class Post(
...
     val creatorId: ObjectId,
     val content: String,
-    val createdAt: LocalDateTime,
-    val thumbsUp: Int
-)
+    val createdAt: LocalDateTime = LocalDateTime.now(),
+    val thumbsUp: Int = 0
+){
+    companion object {
+        const val COLLECTION_NAME = "post"
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/entity/Post.kt:19]: use the const in the Document annotation:
```
@Document(COLLECTION_NAME)
```

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/nats/user/controller/DeleteUserByIdNatsController.kt [hunk 5]
```diff
@@ -18,12 +20,16 @@ class DeleteUserByIdNatsController(
...
-        buildSuccessResponse()
-    }.getOrElse { exception ->
-        buildFailureResponse(exception.javaClass.simpleName, exception.toString())
+        return userService.deleteUserById(id)
+            .then(buildSuccessResponse().toMono())
+            .onErrorResume { ex ->
+                buildFailureResponse(
+                    ex.javaClass.simpleName,
+                    ex.toString()
+                ).toMono() }
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/nats/user/controller/DeleteUserByIdNatsController.kt]:
```suggestion
                ).toMono() 
            }
```

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/nats/user/controller/DeleteUserByUsernameNatsController.kt [hunk 6]
```diff
@@ -18,12 +20,16 @@ class DeleteUserByUsernameNatsController(
...
-        buildSuccessResponse()
-    }.getOrElse { exception ->
-        buildFailureResponse(exception.javaClass.simpleName, exception.toString())
+        return userService.deleteUserByUsername(username)
+            .then(buildSuccessResponse().toMono())
+            .onErrorResume { ex ->
+                buildFailureResponse(
+                    ex.javaClass.simpleName,
+                    ex.toString()
+                ).toMono() }
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/nats/user/controller/DeleteUserByUsernameNatsController.kt]:
```suggestion
                ).toMono() 
            }
```

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/nats/user/controller/CreateUserNatsController.kt [hunk 7]
```diff
@@ -1,33 +1,40 @@
...
-        val savedUser = userService.createUser(request.user.toRequest())
-
-        buildSuccessResponse(savedUser.toProto())
-    }.getOrElse { exception ->
-        buildFailureResponse(exception.javaClass.simpleName, exception.toString())
+    override fun handleHelper(request: CreateUserRequest): Mono<CreateUserResponse> {
+        val user = request.user.toEntity()
+
+        return userService.createUser(user)
+            .map {buildSuccessResponse(it.toProto())}
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/nats/user/controller/CreateUserNatsController.kt]:
```suggestion
            .map { buildSuccessResponse(it.toProto()) }
```

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/nats/user/controller/GetAllUsersNatsController.kt [hunk 8]
```diff
@@ -20,11 +22,16 @@ class GetAllUsersNatsController(
...
 
-    override fun handle(request: GetAllUsersRequest): GetAllUsersResponse = runCatching {
-
-        buildSuccessResponse(userService.getAllUsers().map { it.toProto() })
-    }.getOrElse { exception ->
-        buildFailureResponse(exception.javaClass.simpleName, exception.toString())
+    override fun handleHelper(request: GetAllUsersRequest): Mono<GetAllUsersResponse> {
+        return userService.getAllUsers()
+            .collectList()
+            .map {buildSuccessResponse(it.map { user -> user.toProto() })}
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/nats/user/controller/GetAllUsersNatsController.kt]:
```suggestion
            .map {buildSuccessResponse( it.map { user -> user.toProto() }) }
```

Comment (ajax-gnatiuk-s) [reply listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/nats/user/controller/GetAllUsersNatsController.kt]: Discuss with @OleksandRomaniuk about using `reduce` operator

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/nats/user/controller/GetUserByIdNatsController.kt [hunk 9]
```diff
@@ -20,11 +22,17 @@ class GetUserByIdNatsController(
...
 
-    override fun handle(request: GetUserByIdRequest): GetUserByIdResponse = runCatching {
+    override fun handleHelper(request: GetUserByIdRequest): Mono<GetUserByIdResponse> {
+        val id = request.userId
 
-        buildSuccessResponse(userService.getUserById(request.userId).toProto())
-    }.getOrElse { exception ->
-        buildFailureResponse(exception.javaClass.simpleName, exception.toString())
+        return userService.getUserById(id)
+            .map {buildSuccessResponse(it.toProto())}
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/nats/user/controller/GetUserByIdNatsController.kt]:
```suggestion
            .map { buildSuccessResponse(it.toProto()) }
```

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/nats/user/controller/UpdateUserNatsController.kt [hunk 10]
```diff
@@ -22,14 +24,18 @@ class UpdateUserNatsController(
...
         if (request.user.id == null)
-            throw EntityNotFoundException("There is no id passed")
-        val savedUser = userService.updateUser(request.user.id, request.user.toEntity())
+            throw NotFoundException("There is no id passed")
 
-        buildSuccessResponse(savedUser.toProto())
-    }.getOrElse { exception ->
-        buildFailureResponse(exception.javaClass.simpleName, exception.toString())
+        return userService.updateUser(request.user.id, request.user.toEntity())
+            .map {buildSuccessResponse(it.toProto())}
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/nats/user/controller/UpdateUserNatsController.kt]:
```suggestion
            .map { buildSuccessResponse(it.toProto()) }
```

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/repository/CustomPostRepositoryImpl.kt [hunk 11]
```diff
@@ -3,64 +3,68 @@ package com.nikitahohulia.listeningplatform.repository
...
-        return mongoTemplate.findAll(Post::class.java, collectionName)
+    override fun findAll(): Flux<Post> {
+        return mongoTemplate.findAll(Post::class.java, Post.COLLECTION_NAME)
     }
 
-    override fun findAllByCreatorId(id: ObjectId): List<Post> {
+    override fun findAllByCreatorId(id: ObjectId): Flux<Post> {
         val query = Query().addCriteria(Criteria.where("creatorId").`is`(id))
-        return mongoTemplate.find(query, Post::class.java, collectionName)
+        return mongoTemplate.find(query, Post::class.java, Post.COLLECTION_NAME)
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/repository/CustomPostRepositoryImpl.kt]:
```suggestion
        return mongoTemplate.find(query, Post::class.java)
```

If you declare the collection name inside the annotation `@Document` then passing the collection name to mongoTemplate is not required

Comment (ajax-gnatiuk-s) [reply listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/repository/CustomPostRepositoryImpl.kt]:
```suggestion
        return mongoTemplate.find<Post>(query)
```
Also, your code could be simplified

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/PublisherServiceImpl.kt [hunk 12]
```diff
@@ -1,78 +1,79 @@
...
     }
 
-    override fun getPublisherByPublisherName(publisherName: String): PublisherDtoResponse {
-        val publisher = publisherRepository.findByPublisherName(publisherName)
-            ?: throw NotFoundException("Publisher not found with given publisherName = $publisherName")
-        return publisher.toResponse()
+    override fun getPublisherByPublisherName(publisherName: String): Mono<Publisher> {
+        return publisherRepository.findByPublisherName(publisherName)
+            .switchIfEmpty { NotFoundException("Publisher not found with given publisherName = $publisherName")
+                .toMono() }
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/PublisherServiceImpl.kt]:
```suggestion
            .switchIfEmpty { 
                NotFoundException("Publisher not found with given publisherName = $publisherName").toMono() 
            }
```

Comment (ajax-gnatiuk-s) [reply listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/PublisherServiceImpl.kt]: or
```suggestion
            .switchIfEmpty { 
                NotFoundException("Publisher not found with given publisherName = $publisherName")
                        .toMono() 
            }
```

**Pull request merged**